peer.go

```diff
@@ -159,11 +159,67 @@ func (l *PeerList) Copy() map[string]*Peer {
 
 	listCopy := make(map[string]*Peer)
 	for k, v := range l.peersByHostPort {
-		listCopy[k] = v
+		listCopy[k] = v.Peer
 	}
 	return listCopy
 }
 
+// exists checks if a hostport exists in the peer list.
+func (l *PeerList) exists(hostPort string) (*peerScore, bool) {
+	l.RLock()
+	ps, ok := l.peersByHostPort[hostPort]
+	l.RUnlock()
+
+	return ps, ok
+}
+
+// UpdatePeerHeap updates the peer heap.
+func (l *PeerList) UpdatePeerHeap(p *Peer) {
+	ps, ok := l.exists(p.hostPort)
+	if !ok {
+		return
+	}
+
+	newScore := l.scoreCalculator.GetScore(p)
+	if newScore == ps.readScore() {
+		return
+	}
+
+	l.Lock()
+	ps.setScore(newScore)
+	l.peerHeap.UpdatePeer(ps)
+	l.Unlock()
+}
+
+type peerScore struct {
+	sync.RWMutex
+	*Peer
+
+	score uint64
+	index int
+	order uint64
+}
+
+func newPeerScore(p *Peer, score uint64) *peerScore {
+	return &peerScore{
+		Peer:  p,
+		score: score,
+		index: -1,
+	}
+}
+
+func (ps *peerScore) readScore() uint64 {
+	ps.RLock()
```

Comment:
defer is a little expensive (https://github.com/prashantv/go-bench/blob/master/mutex_results.md), it adds ~80ns. A simple read will be super fast so the defer is going to slow this method, so remove it.

change this code to do
```go
  ps.RLock()
  score := ps.score
  ps.RUnlock()

  return score
```